Make accessible toolbar stable and deprecate old usage

diegohaz opened the pull request:
Closes #18619

Summary of changes in this PR:

- Renamed `__experimentalToolbarItem` component to `ToolbarItem`.
- Renamed `Toolbar`'s `__experimentalAccessibilityLabel` prop to `label`.
- Deprecated `Toolbar` when used without the `label` prop (will still fallback to `ToolbarGroup` for backward compatibility).
- Wrote developer documentation for `Toolbar`, `ToolbarGroup`, `ToolbarButton` and `ToolbarItem` components.

### How to test?

Make sure block toolbar and the top toolbar are working.

diegohaz committed:
Deprecate legacy toolbar usage

diegohaz committed:
Update e2e tests

gziolo commented:
Cross-linking my comment https://github.com/WordPress/gutenberg/issues/23615#issuecomment-653000479 in the tracking issue for experimental APIs.

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

youknowriad commented on packages/block-editor/src/components/navigable-toolbar/index.js:
```diff
@@ -43,7 +44,13 @@ function useIsAccessibleToolbar( ref ) {
-		setIsAccessibleToolbar( hasOnlyToolbarItem( tabbables ) );
+		const onlyToolbarItem = hasOnlyToolbarItem( tabbables );
+		if ( ! onlyToolbarItem ) {
+			deprecated( 'Using custom components as toolbar controls', {
```
Does this mean we should make the `__experimentalAccessibilityLabel` prop stable in this PR too, since that's the alternative right?

diegohaz replied:
Yes!

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

diegohaz committed:
Remove __experimental flag from Toolbar accessibilityLabel prop

youknowriad commented on packages/block-editor/src/components/navigable-toolbar/index.js:
```diff
@@ -90,7 +97,7 @@ function NavigableToolbar( { children, focusOnMount, ...props } ) {
 		return (
 			<Toolbar
-				__experimentalAccessibilityLabel={ props[ 'aria-label' ] }
+				accessibilityLabel={ props[ 'aria-label' ] }
```
Do you think this could be just "label"? what's the reasoning for using "accessibilityLabel"?

diegohaz replied:
This is so it matches the same name as the [React Native prop](https://reactnative.dev/docs/accessibility#accessibilitylabel), but I have no strong opinions on this. I'm totally open for naming it `label` as this is already used by other components like `Button`. cc @gziolo

youknowriad replied:
Unfortunately, @gziolo is away for a few months :P. We'll have to make the decision ourselves. My comment was actually driven by the fact that we already use "label" in other places.

diegohaz replied:
Ah, let's go with `label` then. It's better to be consistent with other components.

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

diegohaz committed:
Use label instead of accessibilityLabel

diegohaz committed:
Deprecate <Toolbar> without label prop

diegohaz committed:
Make ToolbarItem component stable

diegohaz committed:
Deprecate ToolbarButton usage without a parent Toolbar

diegohaz committed:
Simplify jsdocs

diegohaz committed:
data-experimental-toolbar-item to data-toolbar-item

diegohaz committed:
Remove deprecation warning from ToolbarButton

This is actually an invalid warning as it may still be rendered inside <ToolbarGroup />

diegohaz committed:
Update unit test

diegohaz committed:
Update Toolbar README

diegohaz committed:
Update tests

diegohaz committed:
Remove ToolbarGroup test from Toolbar

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

diegohaz committed:
Replace Toolbar by ToolbarGroup on the Classic Editor

diegohaz committed:
Merge branch 'master' into try/accessible-toolbar

diegohaz committed:
Add Toolbar, ToolbarButton, ToolbarGroup and ToolbarItem docs

diegohaz committed:
Add links to depreaction notices and warnings

diegohaz committed:
Update ToolbarButton docs

diegohaz commented:
This is ready for review! :)

youknowriad reviewed (approved):
This is looking great for me. 

I'd like to ensure that I understand the block author story better here (as I think it's important to include in the dev note). How does a block author add an accessible toolbar/buttons to its BlockControls?

If I'm understanding properly, there's no assumptation that a parent Toolbar exists, so we have to add a `Toolbar` with its children, right? And if we want to add a single button, we just use `Button` directly, right?

diegohaz committed:
Improve documentation around BlockControls usage

diegohaz committed:
Update deprecation warning link on Navigable Toolbar

diegohaz committed:
Update ToolbarItem README

diegohaz commented:
@youknowriad Not exactly. I updated the docs here so they're (hopefully) more clear regarding the usage of `BlockControls`.

Basically you have to use either `ToolbarItem` or `ToolbarButton` components in `BlockControls`. Optionally wrapping them with `ToolbarGroup` if you want them in a separate group in the block toolbar.

<a href="https://github.com/WordPress/gutenberg/blob/try/accessible-toolbar/packages/components/src/toolbar-item/README.md#inside-blockcontrols"><img src="https://user-images.githubusercontent.com/3068563/89816011-24baa380-db1c-11ea-808c-75b57501998b.png" alt="documentation screenshot"></a>

<a href="https://github.com/WordPress/gutenberg/blob/try/accessible-toolbar/packages/components/src/toolbar-button/README.md#inside-blockcontrols"><img src="https://user-images.githubusercontent.com/3068563/89816129-5cc1e680-db1c-11ea-9004-aa2ea53dd2a1.png" alt="documentation screenshot"></a>

If you use `Button` or any other tabbable element that is not `ToolbarItem` or `ToolbarButton` inside `BlockControls`, you'll see this warning that will point you to the docs above:

https://github.com/WordPress/gutenberg/blob/e177fb189ee862094b21d7ba98ea1b97ec3763cb/packages/block-editor/src/components/navigable-toolbar/index.js#L65-L69

diegohaz committed:
Update ToolbarItem README

youknowriad commented on packages/components/src/toolbar-item/README.md:
````diff
@@ -47,6 +47,26 @@ function MyToolbar() {
+```jsx
+import { BlockControls } from '@wordpress/block-editor';
+import { Toolbar, ToolbarGroup, ToolbarItem, Button } from '@wordpress/components';
+import { edit } from '@wordpress/icons';
````
Seems like Toolbar and  edit are not used

nerrad commented:
@diegohaz this new API and deprecation wasn't added to the `CHANGELOG.md` for the `@wordpress/components` package. Will you please add details about this to the changelog in the appropriate version section it was released with? Thanks!

 cc @gziolo for visibility.
